### PR TITLE
Thread.sleep() method is replaced with Awaitility

### PR DESCRIPTION
**Pull request by shreyanshR7**
#7154@nastra

### CODE REVIEW

File: api/src/test/java/org/apache/iceberg/metrics/TestDefaultTimer.java [hunk 1]
```diff
@@ -104,7 +106,7 @@ public void measureRunnable() {
     Runnable runnable =
         () -> {
           try {
-            Thread.sleep(100);
+            Awaitility.await().atLeast(Duration.ofMillis(100)).until(() -> true);
```
Comment (nk1506) [inline api/src/test/java/org/apache/iceberg/metrics/TestDefaultTimer.java:109]: won't it become flaky, if callable condition is evaluated sooner. 
`if (evaluationDuration.compareTo(minWaitTime) < 0) {
                String message = String.format("Condition was evaluated in %s which is earlier than expected minimum timeout %s",
                        formatAsString(evaluationDuration), formatAsString(minWaitTime));
                conditionEvaluationHandler.handleTimeout(message, true);
                throw new ConditionTimeoutException(message);
            }`
            
            Why not use something like `https://github.com/apache/iceberg/blob/master/api/src/test/java/org/apache/iceberg/TestHelpers.java#L57`
            ?

**Review by nk1506 — COMMENTED** [review]
IMHO, we should use https://github.com/apache/iceberg/blob/master/api/src/test/java/org/apache/iceberg/TestHelpers.java#L57
Or create a utility with parameter as duration where callable is always returning true.

**Comment by shreyanshR7** [comment]
Oh i see, the code uses a while loop checks the current time until the condition is met.But its asked to replace Thread.sleep method with awaitility, should i implement your suggestion as it don't uses awaitility.@nk1506

File: flink/v1.15/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceContinuous.java [hunk 2]
```diff
@@ -325,7 +328,7 @@ public void testSpecificSnapshotTimestamp() throws Exception {
     long snapshot0Timestamp = tableResource.table().currentSnapshot().timestampMillis();
 
     // sleep for 2 ms to make sure snapshot1 has a higher timestamp value
-    Thread.sleep(2);
+    Awaitility.await().atLeast(Duration.ofMillis(2)).until(() -> true);
```
Comment (nk1506) [inline flink/v1.15/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceContinuous.java:331]: If thread just requires to sleep for some duration , ideally it should be like 
`Awaitility.await().pollDelay(2, TimeUnit.MILLISECONDS).until(() -> System.currentTimeMillis()-snapshot0Timestamp>2);`

File: flink/v1.15/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceContinuous.java [hunk 3]
```diff
@@ -403,7 +406,7 @@ public static List<Row> waitForResult(CloseableIterator<Row> iter, int limit) {
 
   public static void waitUntilJobIsRunning(ClusterClient<?> client) throws Exception {
```
Comment (nk1506) [inline flink/v1.15/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceContinuous.java:407]: remove this method and instead of this it should be something like
`Awaitility.await().pollDelay(10, TimeUnit.MILLISECONDS).until(() -> !getRunningJobs(MINI_CLUSTER_RESOURCE.getClusterClient()).isEmpty());`

**Review by nk1506 — CHANGES_REQUESTED** [review]
I think idea here is to replace all the thread usage with `Awaitility`. But we need to understand why the thread is sleeping first. 

1. If there is intentional sleep to add some delay. It should be handled with corresponding poll-delay and timestamp diff check. 
2. If there is a sleep for another process to be finished, it should check for the corresponding job. 
Make sure `pollDelay` , `atLeast`, `atMost` are being configured correctly.

**Comment by shreyanshR7** [comment]
Thanks for the suggestion @nk1506 ,I'll update that

**Comment by shreyanshR7** [comment]
I've made the changes@nk1506

**Comment by nastra** [comment]
The goal of https://github.com/apache/iceberg/issues/7154 is to convert `Thread.sleep` usages to Awaitility where it makes sense. We don't want to blindly just replace all `Thread.sleep` usage. We need to understand why the sleep is used in the first place.

Typically, `Thread.sleep` is used to wait for an async condition to happen. Rather than waiting the max sleep time, we want to wait less than that, by checking a success condition that indicates we can proceed.

Here is a good example where the test was made more stable by the use of Awaitility: https://github.com/apache/iceberg/blob/cff2ff9d2f533c8e733324a66f3101eafbc2e932/core/src/test/java/org/apache/iceberg/actions/TestCommitService.java#L99-L102

We wait at most 5 seconds until the given assertion is met. 

For this PR that means a good starting point might be https://github.com/apache/iceberg/blob/f9f85e7f2647020ef7e9571f76027c8e55075690/flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/source/TestStreamingMonitorFunction.java#L153, where there's a 1s sleep and then a check. That's the kind of things we want to replace with Awaitility

**Comment by shreyanshR7** [comment]
I've made the changes :)@nastra

File: api/src/test/java/org/apache/iceberg/metrics/TestDefaultTimer.java [hunk 4]
```diff
@@ -104,7 +104,7 @@ public void measureRunnable() {
     Runnable runnable =
         () -> {
           try {
-            Thread.sleep(100);
+           Thread.sleep(100);
```
Comment (nastra) [inline api/src/test/java/org/apache/iceberg/metrics/TestDefaultTimer.java:107]: unnecessary changes that need to be reverted, same for the other files

**Review by nastra — CHANGES_REQUESTED** [review]

File: core/src/test/java/org/apache/iceberg/util/TestInMemoryLockManager.java [hunk 5]
```diff
@@ -112,7 +114,7 @@ public void testAcquireSingleProcess() throws Exception {
     CompletableFuture.supplyAsync(
         () -> {
           try {
-            Thread.sleep(200);
+            Awaitility.await().atLeast(Duration.ofMillis(200)).until(() -> true);
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/util/TestInMemoryLockManager.java:117]: does it make sense here to switch to awaitility?

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceFailover.java [hunk 6]
```diff
@@ -208,7 +210,7 @@ private void testContinuousIcebergSource(FailoverType failoverType) throws Excep
     JobID jobId = jobClient.getJobID();
 
     for (int i = 1; i < 5; i++) {
-      Thread.sleep(10);
+      Awaitility.await().atLeast(Duration.ofMillis(10)).until(() -> true);
```
Comment (nastra) [inline flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceFailover.java:213]: does it make sense here to switch to awaitility?

**Comment by shreyanshR7** [comment]
wait let me check....

**Comment by shreyanshR7** [comment]
I tried to fix the issues :) @nastra

**Comment by nastra** [comment]
@shreyanshR7 there are still many unrelated changes to files. Can you please fix all of those?

**Comment by shreyanshR7** [comment]
@nastra I made another pull request addressing the issue with some more clean changes, please consider that.
